saleor/graphql/order/mutations/draft_orders.py

```diff
@@ -81,16 +87,60 @@ class Meta:
         error_type_class = OrderError
         error_type_field = "order_errors"
 
+    @classmethod
+    def validate_product_is_published_in_channel(
+        cls, info, instance, variants, channel_id
+    ):
+        channel = instance.channel_id
+        if not channel and channel_id:
+            channel = cls.get_node_or_error(info, channel_id, only_type=Channel).id
+
+        if channel:
+            variant_ids = [variant.id for variant in variants]
+            unpublished_variants = ProductChannelListing.objects.filter(
+                Q(product__variants__id__in=variant_ids),
+                Q(channel__id=channel),
+                Q(is_published=False),
+            ).values_list("product__variants__id", flat=True)
+            if unpublished_variants:
+                raise ValidationError(
+                    {
+                        "lines": ValidationError(
+                            "Can't add product variant that are not published in "
+                            "the channel associated with this order.",
+                            code=OrderErrorCode.PRODUCT_NOT_PUBLISHED,
+                            params={"variants": unpublished_variants},
```

Comment:
This should be a list of GraphQL IDs (base64 encoded). Otherwise (when returning database ID) client won't know how to handle them (in Apollo cache we identify objects by GraphQL ID).